src/context/WalletProvider/WalletConnectV2/useWalletConnectV2EventHandler.ts

```diff
@@ -0,0 +1,34 @@
+import type EthereumProvider from '@walletconnect/ethereum-provider'
+import type { Dispatch } from 'react'
+import { useCallback, useEffect } from 'react'
+import type { ActionTypes } from 'context/WalletProvider/actions'
+import { WalletActions } from 'context/WalletProvider/actions'
+import { clearLocalWallet } from 'context/WalletProvider/local-wallet'
+import type { InitialState } from 'context/WalletProvider/WalletProvider'
+
+export const useWalletConnectV2EventHandler = (
+  state: InitialState,
+  dispatch: Dispatch<ActionTypes>,
+) => {
+  const { provider } = state
+  const ethereumProvider = provider as EthereumProvider
+
+  const handleDisconnect = useCallback(() => {
+    /**
+     * in case of KeepKey placeholder wallet,
+     * the disconnect function is undefined
+     */
+    state.wallet?.disconnect?.()
+    dispatch({ type: WalletActions.RESET_STATE })
+    clearLocalWallet()
+  }, [dispatch, state.wallet])
+
+  useEffect(() => {
```

Comment:
Can we add the same guards as this guy? 🙏🏽 

Event handlers running on any keyring (or in this case, any Ethereum provider) are introducing a bunch of bugs

https://github.com/shapeshift/web/pull/5336/files#diff-850c3685f1d3ca0096c3204a6bbf59b35b1780e11e57057d75eec034be4206b9R33-R37

Reply:
🫡 https://github.com/shapeshift/web/pull/5315/commits/f8095de7c2fa7433f7a2c6493e0c1e23c625c0d0